src/reducers/nft/index.js

```diff
@@ -134,15 +166,46 @@ const selectOwnedTokensForAccount = createSelector(
 
 const selectOwnedTokensForAccountForContract = createSelector(
     [selectOwnedTokensForAccount, getContractNameParam],
-    (ownedTokensByContractName, contractName) => ownedTokensByContractName[contractName] || []
+    (ownedTokensByContractName, contractName) => ({
+        ...initialOwnedTokenState,
+        ...ownedTokensByContractName[contractName]
+    })
+);
+
+const selectTokensListForAccountForContract = createSelector(
+    selectOwnedTokensForAccountForContract,
+    (ownedTokensByAccountByContract) => ownedTokensByAccountByContract.tokens
+);
+
+// returns owned tokens for an account containing token array, only if tokens exist
+const selectOwnedTokensForAccountTokenList = createSelector(
+    selectOwnedTokensForAccount,
+    (ownedTokensByContractName) => Object.entries(ownedTokensByContractName)
+        .reduce((x, [contractName, { tokens }]) => !!tokens?.length
```

Comment:
Yes, I was aware of this problem when fixing it, just thought that the new selector will solve the problem and it makes sense to filter only when tokens are already available. But you are right, we shouldn't show tokens if we don't have metadata so your solution is better. Thanks.

Reply:
Any time we run into undefined behaviour, we need to track down the _root cause and fix it_.  If we add little work arounds when we discover broken lifecycles or store data integrity problems we're going to be fighting that problem again and again in the future because the data lifecycle itself is inherently broken.  And often the person who runs into the problem in the future won't be the same person who applied the hack to work around the broken state the first time through; it destroys team velocity and fills the codebase with confusing imperative patches that hide underlying bugs in the implementation for specific use-cases but leave the underlying bugs there to be encountered again later.

Reply:
Of course, I agree, but that's not the case here. I tracked the problem and fixed it by introducing the selector that will return the proper values, which is the tokens array instead of the object that includes (or not includes) the tokens array. So it's not really a "work arounds" :) 

IMO we should use your solution but also use `selectOwnedTokensForAccountTokenList` selector that I prepared. 

Please take a look at the `selectTokensWithMetadataForAccountId` selector, right now this selector is first receiving `ownedTokensByContractName` that have only `loading` state, and after that the it's receiving `ownedTokensByContractName` that includes `tokens` state. So at first the `selectTokensWithMetadataForAccountId` is returning `ownedTokensMetadata` that is undefined - we don't have any use of this selector if the `ownedTokensMetadata` state is `undefined`, so it's unnecesarrly memoized.

 If we use `selectOwnedTokensForAccountTokenList` in `selectTokensWithMetadataForAccountId` then `selectTokensWithMetadataForAccountId` will receive empty array untill the tokens are loaded.

So using `selectOwnedTokensForAccountTokenList` will be more effective, because `selectTokensWithMetadataForAccountId` selector will be memoized few times less.